acb_poly/graeffe_transform.c

```diff
@@ -0,0 +1,52 @@
+/*
+    Copyright (C) 2021 Matthias Gessinger
+
+    This file is part of Arb.
+
+    Arb is free software: you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License (LGPL) as published
+    by the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.  See <http://www.gnu.org/licenses/>.
+*/
+
+#include "acb_poly.h"
+
+void
+_acb_poly_graeffe_transform(acb_ptr b, acb_srcptr a, slong len, slong prec)
+{
+    if (len == 0)
+        return;
+
+    slong q, i;
+
+    q = (len - 1) / 2 + 1;
+    acb_ptr pe = _acb_vec_init(q);
+    acb_ptr po = _acb_vec_init(len);
+
+    for (i = len - 1; i >= 0; i--)
+    {
+        if (i % 2 == 0)
+            acb_set(pe + i / 2, a + i);
+        else
+            acb_set(po + i / 2, a + i);
+    }
+
+    _acb_poly_mul(b, po, q, po, q, prec);
```

Comment:
If the polynomial has even degree, then the odd part will be 1 shorter than the even part. Using this exact length will be a small optimization. You can also skip the _acb_poly_shift_left by writing directly to b + 1.